File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.py
from collections import deque

def solution(maps):
    answer = -1
    
    n = len(maps)
    m = len(maps[0])
    visited = [[False] * m for _ in range(n)] # 맵 방문 여부
    visited[0][0] = True # 출발 지점
    
    dq = deque([(0, 0, 1)]) # (x, y, 현재 거리)
    
    directions = [(-1, 0), (1, 0), (0, -1), (0, 1)] # 상하좌우
    
    while dq:
        x, y, dist = dq.popleft()
        
        # 목적지 도달
        if x == n-1 and y == m-1:
            return dist
        
        for i in range(4):
            nx, ny = x + directions[i][0], y + directions[i][1]
            if 0 <= nx < n and 0 <= ny < m:
                if maps[nx][ny] == 1 and not visited[nx][ny]:
                    visited[nx][ny] = True
                    dq.append((nx, ny, dist+1))
    
    return answer

File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.java
import java.util.*;

class Solution {
    public int solution(int[][] maps) {
        int answer = -1;
        
        int n = maps.length, m = maps[0].length;
        boolean[][] visited = new boolean[n][m];
        visited[0][0] = true;
        
        Queue<Integer[]> queue = new LinkedList<>();
        queue.add(new Integer[]{0, 0, 1});
        
        int[][] directions = {{-1, 0}, {1, 0}, {0, -1}, {0, 1}};
        
        while (!queue.isEmpty()) {
            Integer[] curr = queue.poll();
            int x = curr[0], y = curr[1], dist = curr[2];
            
            if (x == n-1 && y == m-1) {
                return dist;
            }
            
            for (int[] direction: directions) {
                int nx = x + direction[0], ny = y + direction[1];
                if (nx >= 0 && nx < n && ny >= 0 && ny < m) {
                    if (maps[nx][ny] == 1 && !visited[nx][ny]) {
                        visited[nx][ny] = true;
                        queue.add(new Integer[]{nx, ny, dist+1});
                    }
                }
            }
        }
        
        return answer;
    }
}